lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,331 @@
+                val disconnectedEndpoints = endpointMap.values.filter { it.status == EndpointStatus.DISCONNECTED }
+
+                if (connectedOutgoing < desiredOutgoingConnections && disconnectedEndpoints.isNotEmpty()) {
+                    val endpointsToConnect = disconnectedEndpoints
+                        .take(desiredOutgoingConnections - connectedOutgoing)
+
+                    log(LogLevel.INFO, "Initiating connection to ${endpointsToConnect.size} endpoints")
+
+                    val updatedMap = ConcurrentHashMap(endpointMap)
+
+                    endpointsToConnect.forEach { endpoint ->
+                        updatedMap[endpoint.endpointId] = endpoint.copy(status = EndpointStatus.CONNECTING, isOutgoing = true)
+                        requestConnection(endpoint.endpointId)
+                    }
+
+                    _endpointStatusFlow.value = updatedMap
+                }
+            }
+        }
+    }
+
+    private fun requestConnection(endpointId: String) {
+        checkClosed()
+        connectionsClient.requestConnection(name, endpointId, connectionLifecycleCallback)
+            .addOnSuccessListener {
+                log(LogLevel.INFO, "Connection request sent to endpoint: $endpointId")
+            }
+            .addOnFailureListener { e ->
+                log(LogLevel.ERROR, "Failed to request connection to endpoint: $endpointId", e)
+                updateEndpointStatus(endpointId, EndpointStatus.DISCONNECTED)
+            }
+    }
+
+    fun sendMessage(endpointId: String, message: String) {
+        checkClosed()
+        val payload = Payload.fromBytes(message.toByteArray(Charsets.UTF_8))
+        connectionsClient.sendPayload(endpointId, payload)
+            .addOnSuccessListener { log(LogLevel.INFO, "Message sent to $endpointId") }
+            .addOnFailureListener { e -> log(LogLevel.ERROR, "Failed to send message to $endpointId", e) }
+    }
+
+    private fun handleBytesPayload(endpointId: String, payload: Payload) {
+        checkClosed()
+        val bytes = payload.asBytes()
+        if (bytes == null) {
+            log(LogLevel.WARNING, "Received null payload from endpoint: $endpointId")
+            return
+        }
+
+        try {
+            messageReceivedListener?.invoke(endpointId, payload)
+        } catch (e: Exception) {
+            log(LogLevel.ERROR, "Error handling payload from $endpointId", e)
+        }
+
+        try {
+            val message = String(bytes, Charsets.UTF_8)
+            log(LogLevel.INFO, "Received message from $endpointId: $message")
+        } catch (e: Exception) {
+            log(LogLevel.ERROR, "Error parsing payload from $endpointId", e)
+        }
+    }
+
+    private fun handleStreamPayload(endpointId: String, payload: Payload) {
+        checkClosed()
+        payload.asStream()?.asInputStream()?.use { inputStream ->
+            val headerBytes = ByteArray(12)
```

Comment:
By convention never use any literal hardcoded like this other than 0, 1, true, and false. This should be defined as a single constant. If something was added this will get bigger. There should be one and only one place that should be changed.

Important principle - DRY - Dont Repeat Yourself / use a single point of truth.